File: code/day_2.jl
# day 2 in Julia

# part 1
# I basically only know what I learned in day 1
input = readlines("input/day2/input.txt")
h = (i[9:end] for i in input if i[1:3]=="for")
d = (i[6:end] for i in input if i[1:3]=="dow")
u = (i[4:end] for i in input if i[1:3]=="up ")

# answer
sum(parse.(Int,h)) * (sum(parse.(Int, d)) - sum(parse.(Int, u)))

# part 2
# tracking [h, a, d]
# baby's first function
function update_position(line, pos)
    dir = split.(line)[1]
    n = parse(Int,split.(line)[2])
    if dir == "down"
        pos[2] = pos[2]+n
    end
    if dir == "up"
        pos[2] = pos[2]-n
    end
    if dir == "forward" 
        pos[1] = pos[1]+n
        pos[3] = pos[3]+ (n*pos[2])
    end
    return pos
end

pos = [0,0,0]
for i in input # global scope what?!? no idea
    global pos = update_position(i, pos)
end
# answer
pos[1]*pos[3]

File: 2021/code/day_2.R
# Day 2 in R


# Input ----
input <- readLines(here::here('input/day2/input.txt'))


# Part 1 instructions ----
# You are driving a sub
# input is your course
# What do you get if you multiply your final horizontal position by your final depth?

# Solve ----
test_input <- c(
  "forward 5",
  "down 5",
  "forward 8",
  "up 3",
  "down 8",
  "forward 2"
)
# test answer is 150
library(dplyr)
solve1 <- function(input){
  totals <- data.frame(raw=input) %>%
    tidyr::separate(raw, c("dir","amt"),sep="\\s") %>%
    group_by(dir) %>%
    summarise(total = sum(as.integer(amt)), .groups = "drop")
  totals$total[totals$dir=="forward"]*(totals$total[totals$dir=="down"]-totals$total[totals$dir=="up"])
}
solve1(test_input)
solve1(input)

# Part 2 ----
# down X increases your aim by X units.
# up X decreases your aim by X units.
# forward X does two things:
#   It increases your horizontal position by X units.
# It increases your depth by your aim multiplied by X.

# Solve ----
# there is probably an elegant two line solution
solve2 <- function(input){
  totals <- data.frame(raw=input) %>%
    tidyr::separate(raw, c("dir","amt"),sep="\\s") %>%
    mutate(amt = as.integer(amt))
  # vector of horiz, depth, aim
  pos <- c(0,0,0)
  for (i in 1:nrow(totals)){
    if(totals[i,]$dir == "down") pos[3] <- pos[3] + totals[i,]$amt
    if(totals[i,]$dir == "up") pos[3] <- pos[3] - totals[i,]$amt
    if(totals[i,]$dir == "forward"){
      pos[1] <- pos[1] + totals[i,]$amt
      pos[2] <- pos[2] + pos[3]*totals[i,]$amt
    }
  }
  pos[1] * pos[2]
}

solve2(test_input)
solve2(input)
